src/data_model/transformations.jl

```diff
@@ -472,3 +472,917 @@ function apply_voltage_angle_difference_bounds!(eng::Dict{String,<:Any}, vad::Re
+    pp_lb_pu::Union{Real,Missing}=missing, 
+    pp_ub_pu::Union{Real,Missing}=missing, 
+    ng_ub_pu::Union{Real,Missing}=missing,  
+)
+    bus_vbase, _ = calc_voltage_bases(data_eng, data_eng["settings"]["vbases_default"])
+
+    for (id,bus) in data_eng["bus"]
+        vbase = bus_vbase[id]
+        terminals = bus["terminals"]
+
+        has_neutral = neutral_terminal in terminals
+        if has_neutral
+            bus["neutral"] = neutral_terminal
+        end
+        bus_phases = intersect(phase_terminals, terminals)
+        has_phases = !isempty(bus_phases)
+        if has_phases
+            bus["phases"] = bus_phases
+        end
+
+        if has_neutral && !ismissing(ng_ub_pu)
+            bus["vm_ng_ub"] = ng_ub_pu*vbase
+        end
+        if has_phases && has_neutral && !ismissing(pn_lb_pu)
+            bus["vm_pn_lb"] = pn_lb_pu*vbase
+        end
+        if has_phases && has_neutral && !ismissing(pn_ub_pu)
+            bus["vm_pn_ub"] = pn_ub_pu*vbase
+        end
+        if has_phases && length(bus_phases)>1 && !ismissing(pp_lb_pu)
+            bus["vm_pp_lb"] = pp_lb_pu*vbase
+        end
+        if has_phases && length(bus_phases)>1 && !ismissing(pp_ub_pu)
+            bus["vm_pp_ub"] = pp_ub_pu*vbase
+        end
+    end
+end
+
+
+"Calculate no-load starting values for all bus-terminals pairs."
+function calc_start_voltage(data_math::Dict{String,Any}; max_iter=Inf, verbose=false, epsilon::Number=1E-3)
+    @assert data_math["data_model"]==MATHEMATICAL
+
+    node_links = Dict(vcat([[((bus["index"],t), []) for t in bus["terminals"]] for (_, bus) in data_math["bus"]]...))
+    for (id, comp) in [data_math["branch"]..., data_math["switch"]...]
+        f_bus   = comp["f_bus"]
+        f_conns = comp["f_connections"]
+        t_bus   = comp["t_bus"]
+        t_conns = comp["t_connections"]
+
+        arcs = []
+        append!(arcs, [((f_bus,f_conns[k]), (t_bus,t_conns[k]), 1.0) for k in 1:length(f_conns)])
+        append!(arcs, [((t_bus,t_conns[k]), (f_bus,f_conns[k]), 1.0) for k in 1:length(f_conns)])
+        for (fr, to, scale) in arcs
+            push!(node_links[fr], (to, scale))
+        end
+    end
+
+    # initialize v_start for all bts to missing
+    v_start = Dict{Tuple{Int,Any},Union{Complex,Missing}}((bus["index"],t)=>missing for (_, bus) in data_math["bus"] for t in bus["terminals"])
+
+    for (_, bus) in data_math["bus"]
+        # set fixed nodes
+        if haskey(bus, "vm") && haskey(bus, "va")
+            for (i,t) in enumerate(bus["terminals"])
+                v_start[(bus["index"],t)] = bus["vm"][i]*exp(im*bus["va"][i])
+            end
+
+        end
+        # set grounded nodes to zero
+        for t in bus["terminals"][bus["grounded"]]
+            v_start[(bus["index"],t)] = 0.0+im*0.0
+        end
+    end
+
+    if verbose
+        progress = round(sum(ismissing.(values(v_start)))/length(v_start)*100, digits=2)
+        println("it. 0:\t$(100-progress)% specified at start")
```

Comment:
I'm not sure println is the way to go for verbose. Is there an easy way to add a progress bar?

Reply:
This is a relic left-over from debugging an initial, inefficient implementation. The current version is so fast that this is not useful anymore. I changed the `println` to `@debug` statements, and removed the `if verbose` conditions and keyword arguments. I am also fine with simply removing it instead if you prefer.